src/ILCompiler.WebAssembly/src/CodeGen/ILToWebAssemblyImporter.cs

```diff
@@ -1007,16 +1039,18 @@ private void HandleCall(MethodDesc callee, ILOpcode opcode = ILOpcode.call)
                 StackEntry toStore = argumentValues[index];
 
                 TypeDesc argType;
-                if (index == 0 && !callee.Signature.IsStatic)
-                {
-                    if(callee.OwningType.IsValueType)
+                if (index == 0 && !signature.IsStatic)
+                {                    
+                    if (opcode == ILOpcode.calli)
+                        argType = GetWellKnownType(WellKnownType.Object);
```

Comment:
If calling instance method on valuetypes, the implied type of `this` would be a byref. I suspect this line might be problematic. If LLVM really needs to know the type, can we just use the type that got passed as the first argument?

Reply:
I tested a ValueType delegate and the value type gets boxed, not loaded as a byref. Changed this to use the incoming type instead. ValueType delegates don't end up with the right this pointer though; I think we have to account for unboxing stubs somehow. I'll file a new issue to track it.

Reply:
> I tested a ValueType delegate and the value type gets boxed, not loaded as a byref.

Yep, closed instance delegates always need to box the `this`. `Calli` will never have to deal with a valuetype `this` within the generated `Invoke` method of delegates.

The general purpose `calli` will have to support this though (as detailed in the page on `calli` in the ECMA spec).